bpython/autocomplete.py

```diff
@@ -335,6 +335,48 @@ def list_attributes(self, obj):
                 return dir(obj)
 
 
+class ArrayItemMembersCompletion(BaseCompletionType):
+
+    def __init__(self, shown_before_tab=True, mode=SIMPLE):
+        self._shown_before_tab = shown_before_tab
+        self.completers = CumulativeCompleter((
+                            StringLiteralAttrCompletion(mode=mode),
+                            AttrCompletion(mode=mode),
+                            ParameterNameCompletion(mode=mode)),
+                            mode=mode)
+
+    def matches(self, cursor_offset, line, **kwargs):
+        if 'locals_' not in kwargs:
+            return None
+        locals_ = kwargs['locals_']
+
+        r = self.locate(cursor_offset, line)
+        if r is None:
+            return None
+        member_part = r[2]
+        _, _, dexpr = lineparts.current_array_with_indexer(cursor_offset, line)
+        try:
+            exec('temp_val_from_array = ' + dexpr, locals_)
```

Comment:
Also I'm happy to make these changes, I've got it open now at can add them
to the merge commit.

On Sat, Nov 28, 2015 at 1:22 PM, Thomas Ballinger <thomasballinger@gmail.com
> wrote:

> Ah got it, that makes sense. Now that it's just AttrCompletion, maybe
> using AttrCompletion.attr_matches would work, or just duplicating that bit
> of logic.
>
> On Sat, Nov 28, 2015 at 1:12 PM, Shawn Axsom <notifications@github.com>
> wrote:
>
>> In bpython/autocomplete.py
>> <https://github.com/bpython/bpython/pull/577#discussion_r46082715>:
>>
>> > +                            AttrCompletion(mode=mode),
>> > +                            ParameterNameCompletion(mode=mode)),
>> > +                            mode=mode)
>> > +
>> > +    def matches(self, cursor_offset, line, **kwargs):
>> > +        if 'locals_' not in kwargs:
>> > +            return None
>> > +        locals_ = kwargs['locals_']
>> > +
>> > +        r = self.locate(cursor_offset, line)
>> > +        if r is None:
>> > +            return None
>> > +        member_part = r[2]
>> > +        _, _, dexpr = lineparts.current_array_with_indexer(cursor_offset, line)
>> > +        try:
>> > +            exec('temp_val_from_array = ' + dexpr, locals_)
>>
>> I wasn't sure how to do it with safe_eval. From my understanding,
>> safe_eval doesn't allow me to set a temp variable dynamically, it will just
>> return a real variable value. I needed a variable name in locals_ that I
>> could reference by string in completers like AttrCompletion, so I could
>> make use of their logic without any changes to their code.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/bpython/bpython/pull/577/files#r46082715>.
>>
>
>